data-structures-for-storage-and-retrieval/skip-lists/skip_list_oc.go

```diff
@@ -0,0 +1,152 @@
+package main
+
+import (
+	"math/rand"
+	"time"
+)
+
+const maxLevel = 10
+const p = 0.5
+
+type skipNode struct {
+	item    Item
+	forward []*skipNode
+}
+
+type skipListOC struct {
+	header *skipNode
+	tail   *skipNode
+	level  int
+}
+
+func newSkipListOC() *skipListOC {
+	header := &skipNode{
+		forward: make([]*skipNode, maxLevel+1),
+	}
+	tail := &skipNode{
+		Item{"zzzzzzz", "zzzzzzzz"},
+		nil,
+	}
+	header.forward[0] = tail
+	return &skipListOC{header, tail, 1}
+}
+
+func (o *skipListOC) Get(key string) (string, bool) {
+	x := o.header
+
+	for i := o.level; i >= 0; i-- {
+		for x.forward[i] != nil && x.forward[i].item.Key < key {
+			x = x.forward[i]
+		}
+	}
+
+	x = x.forward[0]
+	if x.item.Key != key {
+		return "", false
+	}
+
+	return x.item.Value, true
+}
+
+func (o *skipListOC) Put(key, value string) bool {
+	update := make([]*skipNode, maxLevel+1)
+	x := o.header
+
+	for i := o.level; i >= 0; i-- {
+		for x.forward[i] != nil && x.forward[i].item.Key < key {
+			x = x.forward[i]
+		}
+		update[i] = x
+	}
+
+	x = x.forward[0]
+
+	if x.item.Key == key {
+		x.item.Value = value
+		return false
+	}
+
+	level := randLevel()
+	if level > o.level {
+		for i := o.level; i <= level; i++ {
+			update[i] = o.header
+		}
+
+		o.level = level
+	}
+
+	x = &skipNode{
+		Item{key, value},
+		make([]*skipNode, maxLevel+1),
+	}
+
+	for i := 0; i <= level; i++ {
+		x.forward[i] = update[i].forward[i]
+		update[i].forward[i] = x
+	}
+
+	return true
+}
+
+func (o *skipListOC) Delete(key string) bool {
+	update := make([]*skipNode, maxLevel+1)
```

Comment:
Is there anything you can consolidate between `Put` and `Delete`?

Reply:
You're exactly right. The functionality of finding the node immediately to the left of the insertion/deletion point can be extracted to a function. Thanks!